### DISCUSSION TITLE
signUp: Best Workflow

### DISCUSSION THREAD

**Discussion by MrTob** [supabase/supabase#8233] [Questions]
I would like to register the user via a fomular, for this I have thought of the following flow (confirmation of the mail is activated):

1. User enters his data -> Request signUp() is sent
2. User confirms email address

How can I now check if the email address is already in use or if it is not yet confirmed.
The documentation says the following:
  _If  `signUp()` is called for an existing confirmed user:
      If "Enable email confirmations" is enabled on the "Authentication" -> "Settings" page, an obfuscated / fake user object will be returned._

how can i check if the user is already registered or has not confirmed the mail yet, i always get a fake user returned

**Comment by wcooperscott** [comment]
It is likely that, when calling signUp, it will return with an appropriate error code telling you if the account is registered or awaiting verification. It is worth looking at the error codes that can be thrown and shape your code around the given data.

However, if that is not sufficient, you can create accounts via edge functions. I cannot write you a tutorial but I can point you in the right direction. Your edge function would receive input data such as `username`, `email`, `password`, `displayName`, `favoriteColor`, etc... Then:
* Look up account
* * does account exist? (if so, return appropriate error)
* * is account verified? (if so, return appropriate error)
* Create user:
```
supabase.auth.api.createUser({
  email: 'user@email.com',
  password: 'password',
  user_metadata: { name: 'Yoda' }
})
```
* return new user details or appropriate success response


Note:
I do not believe you can simply "look up" an account. If you give your Edge Functions the Supabase client with an the Service Role key and pass, you might be able to do a SQL search from the `auth.user` table, otherwise you will need to create a `view` of the `auth.user` table. If you do create a user view, be sure to apply the appropriate permissions to it so clients/users cannot query the users table as that contains some sensitive information.

**Comment by GaryAustin1** [comment]
You can not easily check without making a server side call.

What I do, is after someone does the signUp, is change to a waiting message if there is no error.  In my case, I'm using OTP, so have a confirm code box,  but it would work for redirect link also.  Then have a message showing: 

_You have signed up for a new account.  Please check for an email to finish this process or [SignIn](url) instead if you have an account already.  Note: you will not get an email if you have an account._  

SignIn is a link to put back up the sign in screen.

It is not great, but it should be a rare occurrence someone tries to signup by mistake.

**Reply by MrTob** [reply]
I understand what you mean
Let's say i disabled email confirmations. Now i  register two times with the same email, 
`const { user, session, error } = await supabase.auth.signUp({
  email: 'example@email.com',
  password: 'example-password',
})`
I get an error: "User already registered" will be returned.
However, when the email confirmation is enabled, I get a fake/mock user. That doesn't make sense, does it?
Thus, with email confirmation enabled, I have no way to check if the user is using an outstanding confirmation or an already used email. Unless I manually save the status of the registration with a trigger or something like that in a table.

**Reply by wcooperscott** [reply]
@MrTob Does the returned user not have an "email_confirmed" (or similar) attribute?

**Reply by GaryAustin1** [reply]
No, there is no way to know if the email exists or not by design...
https://github.com/supabase/supabase-js/issues/296

**Reply by MrTob** [reply]
@wcooperscott do you mean that an attribute should be set in the raw_user_meta_data.  and then set a trigger on auth.users if the mail was confirmed

**Reply by GaryAustin1** [reply]
You can write an rpc/postgres function to see if an email exists by checking the auth.users table (has to be security definer function). But anyone can call it and probe for emails (which is why they blocked any info on signup).
